File: reference/mathexpr/src/evaluator.test.ts
import { describe, test, expect } from "bun:test";
import { evaluate } from "./evaluator";
import { numberLiteral, unaryExpr, binaryExpr } from "./ast-types";

describe("evaluator", () => {
  test("number literal", () => {
    expect(evaluate(numberLiteral(42))).toBe(42);
  });

  test("unary negation", () => {
    expect(evaluate(unaryExpr("-", numberLiteral(5)))).toBe(-5);
  });

  test("addition", () => {
    expect(
      evaluate(binaryExpr("+", numberLiteral(2), numberLiteral(3))),
    ).toBe(5);
  });

  test("subtraction", () => {
    expect(
      evaluate(binaryExpr("-", numberLiteral(10), numberLiteral(4))),
    ).toBe(6);
  });

  test("multiplication", () => {
    expect(
      evaluate(binaryExpr("*", numberLiteral(3), numberLiteral(7))),
    ).toBe(21);
  });

  test("division", () => {
    expect(
      evaluate(binaryExpr("/", numberLiteral(10), numberLiteral(4))),
    ).toBe(2.5);
  });

  test("modulo", () => {
    expect(
      evaluate(binaryExpr("%", numberLiteral(10), numberLiteral(3))),
    ).toBe(1);
  });

  test("power", () => {
    expect(
      evaluate(binaryExpr("**", numberLiteral(2), numberLiteral(10))),
    ).toBe(1024);
  });

  test("division by zero throws", () => {
    expect(() =>
      evaluate(binaryExpr("/", numberLiteral(1), numberLiteral(0))),
    ).toThrow("Division by zero");
  });

  test("modulo by zero throws", () => {
    expect(() =>
      evaluate(binaryExpr("%", numberLiteral(1), numberLiteral(0))),
    ).toThrow("Modulo by zero");
  });

  test("nested expression: (2 + 3) * -4", () => {
    const expr = binaryExpr(
      "*",
      binaryExpr("+", numberLiteral(2), numberLiteral(3)),
      unaryExpr("-", numberLiteral(4)),
    );
    expect(evaluate(expr)).toBe(-20);
  });
});

File: research/experiments/sweep/runs/mathexpr-full-kotlin-sonnet45/src/test/kotlin/mathexpr/EvaluatorTest.kt
package mathexpr

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class EvaluatorTest {
    @Test
    fun `evaluates number literal`() {
        val node = numberLiteral(42.0)
        assertEquals(42.0, evaluate(node))
    }

    @Test
    fun `evaluates decimal number`() {
        val node = numberLiteral(3.14)
        assertEquals(3.14, evaluate(node))
    }

    @Test
    fun `evaluates unary minus`() {
        val node = unaryExpr("-", numberLiteral(5.0))
        assertEquals(-5.0, evaluate(node))
    }

    @Test
    fun `evaluates double unary minus`() {
        val node = unaryExpr("-", unaryExpr("-", numberLiteral(7.0)))
        assertEquals(7.0, evaluate(node))
    }

    @Test
    fun `evaluates addition`() {
        val node = binaryExpr("+", numberLiteral(2.0), numberLiteral(3.0))
        assertEquals(5.0, evaluate(node))
    }

    @Test
    fun `evaluates subtraction`() {
        val node = binaryExpr("-", numberLiteral(10.0), numberLiteral(4.0))
        assertEquals(6.0, evaluate(node))
    }

    @Test
    fun `evaluates multiplication`() {
        val node = binaryExpr("*", numberLiteral(3.0), numberLiteral(7.0))
        assertEquals(21.0, evaluate(node))
    }

    @Test
    fun `evaluates division`() {
        val node = binaryExpr("/", numberLiteral(15.0), numberLiteral(4.0))
        assertEquals(3.75, evaluate(node))
    }

    @Test
    fun `evaluates modulo`() {
        val node = binaryExpr("%", numberLiteral(10.0), numberLiteral(3.0))
        assertEquals(1.0, evaluate(node))
    }

    @Test
    fun `evaluates power`() {
        val node = binaryExpr("**", numberLiteral(2.0), numberLiteral(8.0))
        assertEquals(256.0, evaluate(node))
    }

    @Test
    fun `evaluates nested expression`() {
        val inner = binaryExpr("*", numberLiteral(3.0), numberLiteral(4.0))
        val node = binaryExpr("+", numberLiteral(2.0), inner)
        assertEquals(14.0, evaluate(node))
    }

    @Test
    fun `throws error on division by zero`() {
        val node = binaryExpr("/", numberLiteral(1.0), numberLiteral(0.0))
        val ex = assertFailsWith<ArithmeticException> {
            evaluate(node)
        }
        assert(ex.message!!.contains("Division by zero"))
    }

    @Test
    fun `throws error on modulo by zero`() {
        val node = binaryExpr("%", numberLiteral(5.0), numberLiteral(0.0))
        val ex = assertFailsWith<ArithmeticException> {
            evaluate(node)
        }
        assert(ex.message!!.contains("Modulo by zero"))
    }
}
